ably/event_emitter.go

```diff
@@ -0,0 +1,176 @@
+package ably
+
+import (
+	"runtime/debug"
+	"sync"
+)
+
+type eventEmitter struct {
+	sync.Mutex
+	listeners listenersForEvent
+	logger    Logger
+}
+
+type emitterEvent = interface{}
+type emitterData = interface{}
+
+type listenersForEvent map[emitterEvent]listenerSet
+type listenerSet map[*eventListener]struct{}
+
+type eventListener struct {
+	handler func(emitterData)
+	once    bool
+
+	queueMtx sync.Mutex
+	queue    []emitterData
+}
+
+func (l *eventListener) handle(e emitterData, log Logger) {
+	// The goroutine that finds the queue empty launches a goroutine for emptying it,
+	// ie. process its emitterData and afterwards any emitterData that concurrent
+	// goroutines may have left in the queue in the meantime.
+	//
+	// Other goroutines just enqueue their emitterData.
+	//
+	// This ensures our guarantee that a single handler runs concurrently with
+	// other handlers, but calls to the same handler are sequential and ordered,
+	// while emitting an event doesn't block the emitting goroutine.
+
+	var isBusy bool
+
+	l.queueMtx.Lock()
+	isBusy = len(l.queue) > 0
+	l.queue = append(l.queue, e)
+	l.queueMtx.Unlock()
+
+	if isBusy {
+		return
+	}
+
+	go func() {
+		done := false
+		for !done {
+			l.queueMtx.Lock()
+			e := l.queue[0]
+			l.queueMtx.Unlock()
+
+			safeHandle(e, l.handler, log)
+
+			l.queueMtx.Lock()
+			l.queue = l.queue[1:]
+			done = len(l.queue) == 0
+			l.queueMtx.Unlock()
+		}
+	}()
+}
+
+func safeHandle(e emitterData, handle func(emitterData), log Logger) {
+	defer func() {
+		r := recover()
+		if r == nil {
+			return
+		}
+		log.Printf(LogError, "EventEmitter: panic in event handler: %v\n%s", r, debug.Stack())
+	}()
+
+	handle(e)
+}
+
+func newEventEmitter(logger Logger) *eventEmitter {
+	return &eventEmitter{
+		listeners: listenersForEvent{
+			nil: listenerSet{},
+		},
+		logger: logger,
+	}
+}
+
+// On registers an event listener. If event is nil, all events will trigger
+// the listener. Otherwise, it must be comparable to the eventEmitter's event
+// type, and only events equal to it will trigger the listener.
+//
+// It returns a function to deregister the listener.
```

Comment:
`On` doesn't return a de-registration function in other libraries (e.g. [`ably-ruby`](https://github.com/ably/ably-ruby/blob/master/lib/ably/modules/event_emitter.rb#L52) or [`ably-js`](https://github.com/ably/ably-js/blob/master/common/lib/util/eventemitter.js#L59)), instead supporting de-registration by passing the listener to `Off`, should we do the same here?

Reply:
We can't, since functions aren't comparable in Go. I thought this was the nicer way to work around that limitation. Another would be to have some kind of comparable function wrappers, but that would be just annoying in the general case in which you don't deregister listeners, and then the `Off` method would be made more complicated too.

Reply:
Can't we just track and compare pointers to the functions, rather than the function values themselves?

Reply:
Hmm, how would that work? You mean like this?

```go
myHandler := func(e) { ... }
...On(&myHandler)
...Off(&myHandler)
```

That would work, but especially given you can't take the address of a function literal (`&func() { ... }`), it would be too annoying in my opinion.

Reply:
Hmm yes, I was thinking about storing the pointers internally, but I didn't think about how the user would pass the pointer back in to `Off`, so yes it seems a return value is the simplest thing 👍